File: elixir/lib/easy/bob.ex
defmodule Bob do
  @spec hey(String.t()) :: String.t()
  def hey(input) do
    trimmed_input = String.trim(input)
    reply(yell?(trimmed_input), question?(trimmed_input), empty?(trimmed_input))
  end

  defp yell?(input), do: String.match?(input, ~r/[[:alpha:]]/) && input == String.upcase(input)
  defp question?(input), do: String.ends_with?(input, "?")
  defp empty?(input), do: input == ""

  defp reply(_is_yell, _is_question, true), do: "Fine. Be that way!"
  defp reply(true, true, _is_empty), do: "Calm down, I know what I'm doing!"
  defp reply(true, _is_question, _is_empty), do: "Whoa, chill out!"
  defp reply(_is_yell, true, _is_empty), do: "Sure."
  defp reply(_is_yell, _is_question, _is_empty), do: "Whatever."
end

File: rust/src/easy/bob.rs
fn is_yell(message: &str) -> bool {
    let is_contain_uppercase = message.chars().any(|c| c.is_ascii_uppercase());
    let is_contain_lowercase = message.chars().any(|c| c.is_ascii_lowercase());

    is_contain_uppercase && !is_contain_lowercase
}

pub fn reply(message: &str) -> &str {
    let last_char = message.trim().chars().last();

    match (last_char, is_yell(message)) {
        (Some('?'), false) => "Sure.",
        (Some('?'), true) => "Calm down, I know what I'm doing!",
        (_, true) => "Whoa, chill out!",
        (None, _) => "Fine. Be that way!",
        _ => "Whatever.",
    }
}
